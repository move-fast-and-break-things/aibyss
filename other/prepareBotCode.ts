import type { BotCode } from "~/other/botCodeStore";
import type { WorldState } from "~/other/world";

type PrepareBotCodeArgs = {
  bot: BotCode;
  botInfo: Record<string, { username: string }>;
  state: Pick<WorldState, "bots" | "food" | "width" | "height">;
  prevBotState: WorldState["bots"];
  botApi: string;
};

export default function prepareBotCode({ bot, botInfo, state, prevBotState, botApi }: PrepareBotCodeArgs): string {
  const { code, username } = bot;

  const botObject = [...state.bots.values()].find(b => bot.id === b.botId);
  if (!botObject) {
    throw new Error(`Bot with id ${bot.id} not found in the world`);
  }

  const me = { x: botObject.x, y: botObject.y, radius: botObject.radius, username };
  const otherPlayers = [...state.bots.values()]
    .filter(b => b.botId !== bot.id)
    .map(b => ({
      x: b.x,
      y: b.y,
      radius: b.radius,
      username: botInfo[b.botId]?.username,
    }));
  const food = state.food.map(f => ({ x: f.x, y: f.y, radius: f.radius }));

  const prevOtherPlayers = [...prevBotState.values()]
    .filter(b => b.botId !== bot.id)
    .map(b => ({ x: b.x, y: b.y, radius: b.radius, username: botInfo[b.botId]?.username }));

  const preparedCode = `
global._player = ${JSON.stringify(me)};
global._otherPlayers = ${JSON.stringify(otherPlayers)};
global._food = ${JSON.stringify(food)};
global._previousOtherPlayersState = ${JSON.stringify(prevOtherPlayers)};
global._worldWidth = ${state.width};
global._worldHeight = ${state.height};

${code}

${botApi}
`;

  return preparedCode;
}
