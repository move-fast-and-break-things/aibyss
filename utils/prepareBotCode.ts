import type { BotCode } from "~/utils/botCodeStore";
import type { WorldState } from "~/utils/world";

type PrepareBotCodeArgs = {
  bot: BotCode;
  state: Pick<WorldState, "bots" | "food" | "width" | "height">;
  botApi: string;
};

export default function prepareBotCode({ bot, state, botApi }: PrepareBotCodeArgs): string | undefined {
  const { code } = bot;

  const botObject = [...state.bots.values()].find(b => bot.id === b.botId);
  if (!botObject) {
    // TODO(yurij): handle this better
    console.error(`Bot with id ${bot.id} not found in the world`);
    return;
  }

  const me = { x: botObject.x, y: botObject.y, radius: botObject.radius };
  const otherPlayers = [...state.bots.values()]
    .filter(b => b.botId !== bot.id)
    .map(b => ({ x: b.x, y: b.y, radius: b.radius }));
  const food = state.food.map(f => ({ x: f.x, y: f.y, radius: f.radius }));

  const preparedCode = `
global._player = ${JSON.stringify(me)};
global._otherPlayers = ${JSON.stringify(otherPlayers)};
global._food = ${JSON.stringify(food)};
global._worldWidth = ${state.width};
global._worldHeight = ${state.height};

${code}

${botApi}
`;

  return preparedCode;
}
