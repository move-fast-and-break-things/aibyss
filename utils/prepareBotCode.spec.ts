import { it, expect } from "vitest";
import prepareBotCode from "./prepareBotCode";

it("prepares bot code correctly when there is one bot in the world", () => {
  const bot = {
    id: "1",
    code: "console.log('hello world');",
  };

  const state = {
    bots: new Map([["1", { x: 0, y: 0, radius: 5, id: "1", color: "#00FF00" }]]),
    food: [],
    width: 100,
    height: 100,
  };

  const botApi = "console.log('hello bot api');";

  const preparedCode = prepareBotCode({ bot, state, botApi });

  expect(preparedCode).toMatchSnapshot();
});

it("prepares bot code correctly when there are two bots in the world", () => {
  const bot = {
    id: "1",
    code: "console.log('hello world');",
  };

  const state = {
    bots: new Map([
      ["1", { x: 0, y: 0, radius: 5, id: "1", color: "#00FF00" }],
      ["2", { x: 10, y: 10, radius: 5, id: "2", color: "#00FF00" }],
    ]),
    food: [],
    width: 100,
    height: 100,
  };

  const botApi = "console.log('hello bot api');";

  const preparedCode = prepareBotCode({ bot, state, botApi });

  expect(preparedCode).toMatchSnapshot();
});

it("prepares bot code correctly when there are three bots and some food in the world", () => {
  const bot = {
    id: "1",
    code: "console.log('hello world');",
  };

  const state = {
    bots: new Map([
      ["1", { x: 0, y: 0, radius: 5, id: "1", color: "#00FF00" }],
      ["2", { x: 10, y: 10, radius: 5, id: "2", color: "#00FF00" }],
      ["3", { x: 20, y: 20, radius: 5, id: "3", color: "#00FF00" }],
    ]),
    food: [{ x: 30, y: 30, radius: 5 }, { x: 40, y: 40, radius: 5 }],
    width: 100,
    height: 100,
  };

  const botApi = "console.log('hello bot api');";

  const preparedCode = prepareBotCode({ bot, state, botApi });

  expect(preparedCode).toMatchSnapshot();
});
