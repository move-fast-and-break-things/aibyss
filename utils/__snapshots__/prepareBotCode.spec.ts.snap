// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`prepares bot code correctly when there are three bots and some food in the world 1`] = `
"
global._player = {"x":0,"y":0,"radius":5};
global._otherPlayers = [{"x":10,"y":10,"radius":5},{"x":20,"y":20,"radius":5}];
global._food = [{"x":30,"y":30,"radius":5},{"x":40,"y":40,"radius":5}];
global._worldWidth = 100;
global._worldHeight = 100;

console.log('hello world');

console.log('hello bot api');
"
`;

exports[`prepares bot code correctly when there are two bots in the world 1`] = `
"
global._player = {"x":0,"y":0,"radius":5};
global._otherPlayers = [{"x":10,"y":10,"radius":5}];
global._food = [];
global._worldWidth = 100;
global._worldHeight = 100;

console.log('hello world');

console.log('hello bot api');
"
`;

exports[`prepares bot code correctly when there is one bot in the world 1`] = `
"
global._player = {"x":0,"y":0,"radius":5};
global._otherPlayers = [];
global._food = [];
global._worldWidth = 100;
global._worldHeight = 100;

console.log('hello world');

console.log('hello bot api');
"
`;
